File: calico_containers/tests/unit/utils_test.py
# Copyright 2015 Metaswitch Networks
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import unittest

from mock import patch
from nose_parameterized import parameterized

from calico_ctl import utils


class TestUtils(unittest.TestCase):

    @parameterized.expand([
        ('127.a.0.1', False),
        ('aa:bb::zz', False),
        ('1.2.3.4', True),
        ('1.2.3.0/24', True),
        ('aa:bb::ff', True),
        ('1111:2222:3333:4444:5555:6666:7777:8888', True),
        ('4294967295', False)
    ])
    def test_validate_cidr(self, cidr, expected_result):
        """
        Test validate_cidr function in calico_ctl utils
        """
        # Call method under test
        test_result = utils.validate_cidr(cidr)

        # Assert
        self.assertEqual(expected_result, test_result)

    @parameterized.expand([
        ('1.2.3.4', 4, True),
        ('1.2.3.4', 6, False),
        ('1.2.3.4', 4, True),
        ('1.2.3.0/24', 4, False),
        ('aa:bb::ff', 4, False),
        ('aa:bb::ff', 6, True),
        ('1111:2222:3333:4444:5555:6666:7777:8888', 6, True),
        ('4294967295', 4, True),
        ('5000000000', 4, False)
    ])
    def test_validate_ip(self, ip, version, expected_result):
        """
        Test validate_ip function in calico_ctl utils
        """
        # Call method under test
        test_result = utils.validate_ip(ip, version)

        # Assert
        self.assertEqual(expected_result, test_result)

    @parameterized.expand([
        ('abcdefghijklmnopqrstuvwxyz', True),
        ('0123456789', True),
        ('profile_1', True),
        ('profile-1', True),
        ('profile 1', False),
        ('profile.1', True),
        ('!', False),
        ('@', False),
        ('#', False),
        ('$', False),
        ('%', False),
        ('^', False),
        ('&', False),
        ('*', False),
        ('()', False)
    ])
    def test_validate_characters(self, input_string, expected_result):
        """
        Test validate_characters function in calico_ctl utils
        """
        with patch('sys.exit', autospec=True) as m_sys_exit:
            # Call method under test
            test_result = utils.validate_characters(input_string)

            # Assert expected result
            self.assertEqual(expected_result, test_result)

    @parameterized.expand([
        ('1.2.3.4', False),
        ('abcde', False),
        ('aa:bb::cc:1234', False),
        ('aa::256', False),
        ('aa...bb:256', False),
        ('aa:256', True),
        ('1.2.3.244:256', True),
        ('1.2.a.244:256', True),
        ('-asr:100', False),
        ('asr-:100', False),
        ('asr-temp-test.thr.yes-33:100', True),
        ('asr-temp-test.-thr.yes-33:100', False),
        ('asr-temp-test.thr-.yes-33:100', False),
        ('asr-temp-test.thr-.yes-33:100', False),
        ('validhostname:0', False),
        ('validhostname:65536', False),
        ('validhostname:1', True),
        ('validhostname:65535', True),
        ('#notvalidhostname:65535', False),
        ('verylong' * 100 + ':200', False),
        ('12.256.122.43:aaa', False)
    ])
    def test_validate_hostname_port(self, input_string, expected_result):
        """
        Test validate_hostname_port function.

        This also tests validate_hostname which is invoked from
        validate_hostname_port.
        """
        test_result = utils.validate_hostname_port(input_string)

        # Assert expected result
        self.assertEqual(expected_result, test_result)

File: hack/release/utils_test.go
// Copyright (c) 2025-2026 Tigera, Inc. All rights reserved.

// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import (
	"testing"
)

func TestIsReleaseVersionFormat(t *testing.T) {
	t.Parallel()

	cases := []struct {
		version string
		want    bool
	}{
		{
			version: "v3.25.0",
			want:    true,
		},
		{
			version: "v3.25.0-rc1",
			want:    false,
		},
		{
			version: "v3.25.0-1.0",
			want:    false,
		},
		{
			version: "3.25.0-1.0-rc1",
			want:    false,
		},
		{
			version: "not-a-version",
			want:    false,
		},
	}

	for _, tc := range cases {
		t.Run(tc.version, func(t *testing.T) {
			t.Parallel()
			got, err := isReleaseVersionFormat(tc.version)
			if err != nil {
				t.Fatalf("isReleaseVersionFormat(%q) unexpected error: %v", tc.version, err)
			}
			if got != tc.want {
				t.Fatalf("isReleaseVersionFormat(%q) = %v, want %v", tc.version, got, tc.want)
			}
		})
	}
}

func TestIsEnterpriseReleaseVersionFormat(t *testing.T) {
	t.Parallel()

	cases := []struct {
		version string
		want    bool
	}{
		{
			version: "v3.25.0",
			want:    true,
		},
		{
			version: "v3.25.0-rc1",
			want:    false,
		},
		{
			version: "v3.25.0-1.0",
			want:    true,
		},
		{
			version: "3.25.0-1.0-rc1",
			want:    false,
		},
		{
			version: "not-a-version",
			want:    false,
		},
	}

	for _, tc := range cases {
		t.Run(tc.version, func(t *testing.T) {
			t.Parallel()
			got, err := isEnterpriseReleaseVersionFormat(tc.version)
			if err != nil {
				t.Fatalf("isEnterpriseReleaseVersionFormat(%q) unexpected error: %v", tc.version, err)
			}
			if got != tc.want {
				t.Fatalf("isEnterpriseReleaseVersionFormat(%q) = %v, want %v", tc.version, got, tc.want)
			}
		})
	}
}

func TestAddTrailingSlash(t *testing.T) {
	t.Parallel()
	cases := []struct {
		input    string
		expected string
	}{
		{"docker.io", "docker.io/"},
		{"quay.io/", "quay.io/"},
		{"gcr.io/some-repo", "gcr.io/some-repo/"},
		{"", ""},
	}

	for _, tc := range cases {
		t.Run(tc.input, func(t *testing.T) {
			t.Parallel()
			got := addTrailingSlash(tc.input)
			if got != tc.expected {
				t.Errorf("expected %q, got %q", tc.expected, got)
			}
		})
	}
}

func TestIsPrereleaseEnterpriseVersion(t *testing.T) {
	t.Parallel()

	cases := []struct {
		version string
		want    bool
	}{
		{
			version: "master",
			want:    false,
		},
		{
			version: "release-v3.25",
			want:    false,
		},
		{
			version: "release-calient-v3.25",
			want:    false,
		},
		{
			version: "v3.25.0",
			want:    false,
		},
		{
			version: "v3.25.0-rc1",
			want:    false,
		},
		{
			version: "v3.25.0-1.0",
			want:    true,
		},
		{
			version: "v3.25.0-2.0",
			want:    true,
		},
	}
	for _, tc := range cases {
		t.Run(tc.version, func(t *testing.T) {
			// t.Parallel()
			got, err := isPrereleaseEnterpriseVersion(tc.version)
			if err != nil {
				t.Fatalf("isPrereleaseEnterpriseVersion(%q) unexpected error: %v", tc.version, err)
			}
			if got != tc.want {
				t.Fatalf("isPrereleaseEnterpriseVersion(%q) = %v, want %v", tc.version, got, tc.want)
			}
		})
	}
}
